src/scalarstats.jl

```diff
@@ -261,36 +261,89 @@ realXcY(x::Real, y::Real) = x*y
 realXcY(x::Complex, y::Complex) = real(x)*real(y) + imag(x)*imag(y)
 
 """
-    sem(x)
+    sem(x; mean=nothing)
+    sem(x::AbstractArray[, weights::AbstractWeights]; mean=nothing)
 
-Return the standard error of the mean of collection `x`,
-i.e. `sqrt(var(x, corrected=true) / length(x))`.
+Return the standard error of the mean for a collection `x`. When not using weights, this is
+the (sample) standard deviation divided by the sample size. If weights are used, the 
+variance of the sample mean is calculated as follows:
+
+* `AnalyticWeights`: Not implemented.
+* `FrequencyWeights`: ``\\frac{\\sum_{i=1}^n w_i (x_i - \\bar{x_i})^2}{(\\sum w_i) (\\sum w_i - 1)}``
+* `ProbabilityWeights`: ``\\frac{n}{n-1} \\frac{\\sum_{i=1}^n w_i^2 (x_i - \\bar{x_i})^2}{\\left( \\sum w_i \\right)^2}``
+
+The standard error is then the square root of the above quantities.
+
+# References
+
+Carl-Erik Särndal, Bengt Swensson, Jan Wretman (1992). Model Assisted Survey Sampling.
+New York: Springer. pp. 51-53.
 """
-function sem(x)
+function sem(x; mean=nothing)
+    if isempty(x)
+        # Return the NaN of the type that we would get for a nonempty x
+        return sqrt(var(x; mean=mean, corrected=true) / 0)
```

Comment:
Seems `var` supports empty iterables only in more recent Julia versions but e.g. not in Julia 1.0 (https://github.com/JuliaLang/julia/blob/5d4eaca0c9fa3d555c79dbacdccb9169fdf64b65/stdlib/Statistics/src/Statistics.jl#L150-L152). So I guess similar to the current implementation unfortunately we have to copy the code in `Statistics.var`:
```suggestion
        # Return the NaN of the type that we would get for a nonempty x
        # This is consistent with `var` but takes into account `mean` as well
        T = eltype(x)
        _mean = mean === nothing ? zero(T) / 1 : zero(mean)
        return oftype(sqrt((abs2(zero(T) - _mean) + abs2(zero(T) - _mean) / 2), NaN)
```

Reply:
Whoops, didn't see your response before the latest commit, where I removed the tests about throwing a method error when testing past 1.6.

I think the current behavior is actually correct, and the previous behavior was a bug -- `sem` is supposed to have the same behavior as `var`. In addition, calling `sem` on different representations of the same object could result in contradictory behavior. (Actually, I just realized I should add tests for calling `sem` with weights on empty iterators.)

Reply:
Whoops, didn't see your response before the latest commit, where I removed the tests about throwing a method error when testing past 1.6.

I think the current behavior is actually correct, and the previous behavior was a bug -- `sem` is supposed to have the same behavior as `var`. In addition, calling `sem` on different representations of the same object could result in contradictory behavior. (Actually, I just realized I should add tests for calling `sem` with weights on empty iterators.)

Reply:
@devmotion do you think we should keep the `var/sem` behavior consistent, or keep the bug how it is now?

Reply:
What is the bug? At least we cannot turn something into an error as it would be breaking, but we can turn errors into something else.

BTW, we could drop support for Julia 1.0 as 1.6 is the LTS now.

Reply:
That said, it's weird to rely on `var` in this method as we don't call it when the input is not empty either. So maybe it's cleaner to compute the type ourselves.

Reply:
The bug is that for old versions of Julia before 1.6, the behavior of `sem` doesn't match that of `var` like it's supposed to. In old versions, `var` throws an error for empty collections, rather than returning `NaN`; the current implementation of `sem` didn't test for this, so instead it always returns `NaN`, even when running on old versions.

Reply:
This is not a bug in `sem` but rather one in `var` which was fixed in more recent Julia versions. In any case, as @nalimilan said, we can't start throwing an error in these cases that currently work just fine.